### DISCUSSION TITLE
IfcWorkPlan for Construction Sequence per AASHTO LRFD BDS 2.5.3

### DISCUSSION THREAD

**Discussion by RickBrice** [jwouellette/TPF-5_372-Unit_Test_Suite#16] [General]
I was recently experimenting with IDS to see if I could check if a model conforms to the information requirement defined in AASHTO LRFD Bridge Design Specifications Article 2.5.3. See highlighted text.
![image](https://github.com/user-attachments/assets/9747febf-a9fb-4306-98b9-156f22576b93)

Do we intend to include the construction sequence in the model? If so, how? [IfcWorkPlan](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcWorkPlan.htm) might do the job.

The IfcWorkPlan would be IfcRelAssignsControl with IfcTask entities that define the high-level steps of the construction sequence. The  IfcRelAggregates with IfcWorkSchedule would not be needed, or logical, since the exact schedule of construction is not known at design time.

One problem with IfcWorkPlan is that it is IfcRelDeclared at the IfcProject level. For say an interchange project with multiple bridges of different types (precast concrete, CIP-segmental, spliced girders, etc.) there will be a different construction sequence for each bridge. 

Here is a model (albeit incomplete) that attempts to define the construction sequence with IfcWorkPlan.

[PGSuper1.ifc.txt](https://github.com/user-attachments/files/17991310/PGSuper1.ifc.txt)

The requirements of Article 2.5.3 can be achieved by reference to external documents.

Here are three examples of typical Washington State DOT construction sequences for precast concrete superstructures.
https://wsdot.wa.gov/publications/fulltext/Bridge/Web_BSD/5.6_A2_1.PDF
https://wsdot.wa.gov/publications/fulltext/Bridge/Web_BSD/5.6_A2_2.PDF
https://wsdot.wa.gov/publications/fulltext/Bridge/Web_BSD/5.6_A2_3.PDF

Per the discussion in https://github.com/jwouellette/TPF-5_372-Unit_Test_Suite/discussions/12, there are several options for referencing external documents.

Would a single, well-defined method of exchanging this information best serve the needs of the contractor in the design-to-construction exchange or is this to be open ended with implementation at the modeler's discretion?

**Comment by MatthiasWeise** [comment]
Yes, the IFC process extension can be used to define construction schedules (see https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/ifcprocessextension/content.html). As far as I know, it was last extended for IFC4 but never really part of an official view or software implementation. 

The new A-bV is a bit unclear about the scope as for instance IfcTask and IfcWorkControl is marked to be included, but IfcWorkSchedule, IfcWorkPlan etc. are not included (https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/annex-b1.html).

**Comment by RickBrice** [comment]
As a Software Vendor Engagement group supporting the TPF work, I'm trying to engage to answer the question "Do we intend to include an assumed construction sequence in an IFC model to support the requirements of AASHTO LRFD 2.5.3, and if so, what is that standardized method?" How have others approached satisfying this requirement?

**Comment by TLiebich** [comment]
We had a discussion at our last IFC for Bridges and Structures internal technical meeting about these information requirements. Given that the requirements as we understood is about identifying, whether a series of elements belongs to a construction sequence, than the recommendation would be to define and use a particular property set with properties to store the number(s) of the construction sequence for each element. This would be straight forward and within the boundaries of the IFC4.3 AbV and the IFC support of most software packages.

The IfcWorkplan, IfcTask, IfcRelSequence, etc. are important if a full construction schedule with start-end time, duration, dependencies, repetition, etc. is required. For representing purely a construction sequence, this would be a too big overhead.

So best to standardize the property set as part of the IDM for the exchange.

**Comment by RickBrice** [comment]
> recommendation would be to define and use a particular property set with properties to store the number(s) of the construction sequence for each element

1) Is there a particular property set in mind, or would this be a property set defined in the TPF/USBridge Data Dictionary? It seems like we need a property set to define the assumed construction sequence for IfcBridge and then a different, but related property set, which indicates which step or steps in the assumed construction sequence is assigned to the element.
2) Some aspects of the construction sequence are more akin to contractor actions. An example "install temporary bracing" and "detension temporary strands" for pretensioned members as shown in the sample construction sequences provided above.

Perhaps an example would be something like the following for the assumed construction sequence of a single span pretensioned concrete girder bridge ([https://wsdot.wa.gov/publications/fulltext/Bridge/Web_BSD/5.6_A2_1.PDF](https://wsdot.wa.gov/publications/fulltext/Bridge/Web_BSD/5.6_A2_1.PDF)).

Property Set: TPFBridge_AssumedConstructionSequence
Defintion: Construction sequence assumed during design to induce certain stresses under dead load, per AASHTO LRFD BDS 2.5.3.
Relationshipe: IfcBridge
Property: ConstructionStep
DataType: String
MinOccurences: 1
MaxOccurence: unlimited

Property Set: TPFBridge_AssumedConstructionSequenceOccurence
Definition: Associates element with a step in the assumed construction sequence
Property: Step
DataType: Integer
MinOccurences: 1
MaxOccurences: unlimited

The following instances of the ConstructionStep property would represent Steps 1-4 in the example construction sequence:

1. Construct end piers
2. Install elastomeric bearing pads
3. Erect girders
4. Install temporary bracing for erection in accordance with Std. Spec. Section 6-02.3(17)F4.
5. Remove expanded polystyrene in blockouts in top flange of girders. Once the expanded polystyrene has been removed from the strand detensioning blockout, prevent moisture for entering the blockout until the temporary top strand is cut and the blockout filled with grout
6. Cut strands in blockouts. Strands may be cut by using a cutting torch and moving the flame back and forth over the length of the exposed strand to let individual wires break one at a time to lessen the shock to the girder. Strands shall be released in a symmetrical manner about the girder centerline starting with those furthest from the centerline and working inwards. For post-tensioned temporary top strands, actively restrain the strand chucks at the girder ends during cutting.
7. Within 24 hours of cutting the temporary strands, fill the blockouts with a grout conforming to Std. Spec. 9-20.3(2). Remove all moisture in blockouts prior to filling them with grout.
8. Install temporary bracing for diaphragm and deck placement in accordance with Std. Spec. Section 6-02.3(17)F5.
9. Cast end and intermediate diaphragms
10. Form and place deck reinforcement
11. Cast bridge deck when diaphragm concrete compressive strength has reached 3000 psi (minimum)
12. Cast traffic barrier after the deck concrete compressive strength has reached 3000 psi (minimum)

Then an IfcBearing would be classified with TPFBearing which as the TPFBridge_AssumedConstructionSequenceOccurance and the Step property would be assigned a value of 2 indicating the bearing is assumed to be added to the bridge as defined by the second instance of the ConstructionStep property of the TPFBridge_AssumedConstructionSequence property set of the TPFBridge class.

**Comment by SergejMuhic** [comment]
Would it not make more sense to go for @RickBrice proposed `IfcWorkPlan` approach? Maintaining such information in psets seems like quite the task, looking at the example with the 12 steps. With `IfcWorkPlan` a standardized library could be attached to the spec which can be referenced in instance files.

True, the same approach could be used with a property set and property enums I guess, but why not use the purposely designed data structures? It seems to be a more efficient approach. Every task just takes a line in an instance file and they can be sequenced according to any requirements. Otherwise, a single property can only capture one step in the sequence.

**Comment by RickBrice** [comment]
I've been experimenting with representing the required assumed construction sequence per AASHTO LRFD 2.5.3 and the required detensioning sequence for temporary strands per AASHTO LRFD 5.9.4.5. There is substantial discussion at https://community.osarch.org/discussion/comment/23500.

It seems following the buildingSmart example for construction task schedule is probably the best approach. https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/annex_e/construction-scheduling/construction-scheduling-task.html

It will support projects with different types of bridges with different assumed construction sequences in the same IfcProject.

The basic information structure is
~~~
IfcProject - IfcRelDeclares - IfcWorkPlan - IfcRelAggregates - IfcWorkSchedule - IfcRelAssignsToControl - IfcTask (summary) - IfcRelNests - [Subtasks IfcTask, IfcTask, etc]
                                                             |                                                 |
                                                             |                                        IfcRelAssignsToProduct
                                                             |                                                 |
                                                             |                                             IfcBridge (bridge #1)
                                                             |
                                                             |
                                                             - IfcWorkSchedule - IfcRelAssignsToControl - IfcTask (summary) - IfcRelNests - [Subtasks IfcTask, IfcTask, etc]
                                                                                                               |
                                                                                                      IfcRelAssignsToProduct
                                                                                                               |
                                                                                                           IfcBridge (bridge #2)
~~~

One problem is the required StartTime attribute on IfcWorkPlan and IfcWorkSchedule. The required CreationDate can be "today", whenever the software generates the assumed construction sequence. Since the sequence is assumed and we don't know when the work will start, what to use for the required StartTime attribute? Perhaps using the same date/time as CreationDate is a good option.

Another problem is the question as to whether or not the required entities are part of the AbV. IfcWorkPlan and IfcWorkSchedule documentation both say "This entity is not part of a standardized schema subset or implementation level." However, these are both sub-types of the abstract IfcWorkControl which is explicitly included in the AbV. The Control Assignment and ProductAssignment concept templates are not part of the AbV. IfcControl and IfcObjectDefinition are part of AbV, but IfcRelAssignsToControl is not. Similarly, IfcProduct and IfcObject are AbV, but IfcRelAssignsToProduct is not. It doesn't make sense that Controls, Object Definition, Product and Object are in the AbV but they cannot be related to one another in the AbV. Perhaps we can infer that everything needed to create a Construction Scheduling Task per bSI's example is part of the AbV.

Attached is an example that I came up with.

[PGSuper1.ifc.txt](https://github.com/user-attachments/files/18400098/PGSuper1.ifc.txt)

**Comment by MatthiasWeise** [comment]
Hi Rick and all, 

thank you for the interesting discussion and example! 
I was involved in earlier discussions about the IFC process extension, which led to changes/extensions in the IFC4 schema.  
We struggled with the definition of _IfcWorkPlan_ and _IfcWorkSchedule_ as well, but decided to more or less keep the definition from IFC2x3 (with the exception of a minor harmonisation for predefinedType and time definitions).

My understanding is that _IfcWorkPlan_ and _IfcWorkSchedule_ are mainly for grouping purposes that helps to organize work (including _IfcWorkCalendar_). It is very flexible and allows the definition of more than one level in a hierarchical tree structure. 
Please note that _IfcTask_ is using _IfcRelAssigns(ToControl)_ instead of _IfcRelAggregates_ meaning that you technically could assign a task to several work plans/schedules if needed. 
Also, if you focus on task sequencing you may not need _IfcWorkPlan_ and _IfcWorkSchedule_ as there is a mandatory start time attribute.
 
The IFC documentation is not fully clear about correct use of _IfcTask_ and for me it is up to an implementer agreement whether you should be forced to create the suggested structure with _IfcWorkPlan_ and _IfcWorkSchedule_. 
Technically you do not need it and beside proposing a summary/root task the documentation also suggests to define a _IfcRelDeclares_ relationship between _IfcProject_ and the summary/root task (see concept usage for object nesting). 
This _IfcRelDeclares_ relationship would be superfluous if there is an assignment to _IfcWorkPlan_ or _IfcWorkSchedule_, which is expected for typical construction schedules with start and end dates.  
So, unless you see a need for work plan/schedule you could simplify the proposed structure by linking the summary/root task via _IfcRelDeclares_ directly to _IfcProject_.
Such structure may not be expected by Bonsai, Synchro, Bexel and other tools and thus not properly handled, but I do not know additional implementer agreements for proper implementation of the IFC process extension. 

Regarding the scope of AbV: 
Documentation of IFC4.3 is not clear about it, but I would be supprised if any entity of the process extension is part of the AbV.
Although I am not a fan of agreements based on custom properties, the solution proposed by Thomas should be simple and is within the scope of the AbV. 

Best regards,
Matthias

**Reply by SergejMuhic** [reply]
I am not sure how to read this "From IFC4 onwards it is required to have a summary task (root of all tasks), which is used to define a link to the work plan or work schedule." To me this seems to imply that either `IfcWorkPlan` or `IfcWorkSchedule` would be needed.

In our case, where we have an assignment to bridge, I would think that either the summary task or the work plan / schedule would not need to be declared on the project either, right?

**Comment by RickBrice** [comment]
Great discussion. I think it is imperative to have a decision how the requirements of AASHTO 2.5.3 (and 5.9.4.5) are to be addressed in a model-based design-to-construction workflow. I don't think the IDM or IDS address this. Should assumed construction sequences be model based information or conveyed from the designer to the contractor by some other mechanism (such as 2D drawing, narrative description in contract specifications, other). 

I think including the assumed construction sequence in the model is the way to go.
1) Contractors can update the model with the actual construction sequence for the as-built model (important for understanding locked in stresses)
2) Future designers and load raters can find this information in the model and not have to search for it in some external document. This could also aid in automated structural analysis model generation in FEA packages.
3) Model-based construction sequence could aid in visualization of different stages of construction and accurately convent design intent.

Assuming there is agreement with communicating the assumed construction sequence in the model, the next issue is how to do it. There are many ways as have been discussed. I explored using just a root-level task that is IfcRelDeclares with IfcProject. This approach eliminates the issue with StartTime of IfcWorkPlan and IfcWorkSchedule, but that issue is easily overcome with a "dummy" StartTime. Regardless of the information structure, an IfcProject in general has more than one bridge and each bridge has a unique assumed construction sequence. This requires that each IfcTask (summary task of assumed construction sequence) is related to its IfcBridge. This is accomplished with IfcTask-IfcRelAssignsToProduct-IfcBridge, but IfcRelAssignsToProduct doesn't seem to be part of the AbV. With or without IfcWorkPlan/IfcWorkSchedule, the representation of the construction sequence with IfcTask(s) still needs to be associated with the bridge.

~~~
IfcProject - IfcRelDeclares - IfcTask (summary) - IfcRelNests - [Subtasks IfcTask, IfcTask, etc]
                            |        |
                            |    IfcRelAssignsToProduct
                            |        |
                            |     IfcBridge (bridge #1)
                            |
                            |
                            - IfcTask (summary) - IfcRelNests - [Subtasks IfcTask, IfcTask, etc]
                                     |
                                 IfcRelAssignsToProduct
                                     |
                                  IfcBridge (bridge #2)
~~~

I don't see this information structure being more or less complicated than the one with IfcWorkPlan/IfcWorkSchedule. If software products currently deal with the IfcWorkPlan/IfcWorkScheule approach, why not leverage that instead of using a simplified structure that isn't supported?

Defining custom property sets and associating them with one another with a property that acts like a database key field seems clumsy. Why do we need to invent a method of conveying the information when one already exists (even if it isn't perfect)? IFC needs to be simplified, not made more complex by inventing one-off custom data structures.

Do the bridge design specifications for other countries have similar requirements for communicating the construction sequence assumed in design? If so, how are they dealing with this?

As far as implementation agreements are concerned, how does that process work? I've raised discussions in the Implementers Forum only for them to go nowhere.  The process for agreements seems to be abandoned. Looking at [https://standards.buildingsmart.org/documents/Implementation/IFC_Implementation_Agreements/](https://standards.buildingsmart.org/documents/Implementation/IFC_Implementation_Agreements/) there hasn't been an official agreement in over 10 years and two proposals have been in-progress for a decade. I don't have confidence official implementation agreements can be made.

**Reply by SergejMuhic** [reply]
```
 IfcProject - IfcRelDeclares - IfcTask (summary) - IfcRelNests - [Subtasks IfcTask, IfcTask, etc]
                             |        |
                             |    IfcRelAssignsToProduct
                             |        |
                             |     IfcBridge (bridge #1)
                             |
                             |
                             - IfcTask (summary) - IfcRelNests - [Subtasks IfcTask, IfcTask, etc]
                                      |
                                  IfcRelAssignsToProduct
                                      |
                                   IfcBridge (bridge #2)

```

@RickBrice I have taken this diagram as a basis and started the Task concept root:
![image](https://github.com/user-attachments/assets/ca52dcb6-b92d-41ce-892f-0f43e9ec66b1)

Also generated the empty templates since we have not yet talked about what kind of a check we want. This means I do not have any RuleIDs created apart from the ones that are used to set and check the types. But here a bit of foretaste:
1. the template
![image](https://github.com/user-attachments/assets/fb26538d-d0c2-4200-a30b-124ee25b7c94)
2. the concept
![image](https://github.com/user-attachments/assets/596c1a01-da72-4f13-9725-a203148570a9)

And the resulting code (just the template):
![image](https://github.com/user-attachments/assets/d71161ab-cc48-47ec-9c64-cfc88de0ef8a)

So, now I can test your file @RickBrice 😄

**Reply by SergejMuhic** [reply]
Test is also ready and waiting for "insertRichardsFileHere":
![image](https://github.com/user-attachments/assets/0b42d86e-a39f-41ca-a8f7-262a2cc5901e)

**Reply by RickBrice** [reply]
[PGSuper1.ifc.txt](https://github.com/user-attachments/files/18441407/PGSuper1.ifc.txt)

**Reply by SergejMuhic** [reply]
Alright this took awhile, apologies. Many meetings later and some family time yesterday, I am back at testing the file according to the proposed structure. Really good work. Passes all basic checks, apart from:

`13s6piTaH4duITiwjqo2Ah IfcTask is not declared.`

I continued with the assignment to product, but here I have a question, would the subtasks also have to be assigned to the bridge or would you see the nesting relationship already supply the necessary connection to the bridge. If this is the case, I have to write the rule a bit differently of course because a valid IfcTask can either be a subtask or assigned to bridge. Currently, the test of course fails because not all IfcTasks are assigned to the bridge.

There are many other things to test in the file of course, the declaration to IfcProject (this one I do not have documented yet since I have a basic file validation that already checks this). It probably would make sense to incorporate this into an MVD also.

The other one is Nesting of IfcTask, I also see an assignment to a IfcWorkSchedule that seems missing from your diagram and since we were talking about sequencing, this could also be modelled as
![image](https://github.com/user-attachments/assets/8084a572-81b6-4f07-a902-81f175934a0f)
https://bsi-infraroom.github.io/IFC-Documentation-Tunnel/4_4_0_0/general/HTML/link/sequential-connectivity.htm

The exercise is really good. Let's keep this discussion going because it is producing amazing results. Also improving my modelling competence with things to consider regarding writing actual tests (see above, all tasks not to be tested for assignment to product).

**Reply by MatthiasWeise** [reply]
@SergejMuhic see separate discussion about product assignment to task. I believe this should be an optional relationship. It is also fine to assign on root/summary task level to IfcBridge. If necessary, you can assign subtasks to parts of the bridge (girder, pile etc.), but re-assignment of subtasks to the bridge again does not add much additional information. 

I am not sure if IfcWorkSchedule is needed (technically it could work without a schedule), but you are right with IfcRelSequence for defining the logical dependency of tasks.

**Reply by SergejMuhic** [reply]
Alright, as clarified in our conversation on Skype, I see have made following adjustment:
![image](https://github.com/user-attachments/assets/3cff9d4e-87c5-4e48-a277-fefdc18f0a77)
i. e. inversing the check that will now just look for an assignment of an `IfcTask` on `IfcBridge.`

To cover the entire diagram, I would create a `Project` root that covers the declaration of a `Task` and a `Task` root for the nesting and the sequencing of `Tasks`. @RickBrice  The one thing I am not certain about is whether both `sequencing` and `nesting` is required or are they redundant, meaning either [sequential connectivity](https://bsi-infraroom.github.io/IFC-Documentation-Tunnel/4_4_0_0/general/HTML/schema/ifcprocessextension/lexical/ifctask.htm#sequential-connectivity) or [nesting](https://bsi-infraroom.github.io/IFC-Documentation-Tunnel/4_4_0_0/general/HTML/schema/ifcprocessextension/lexical/ifctask.htm#object-nesting) or both?

**Reply by RickBrice** [reply]
As I understand nesting, it defines an order so it seems it would be enough. In the example we are exploring all the tasks are executed in order without overlap. Sequencing is redundant. I added IfcRelSequence because of this

![](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/figures/ifctask_instantiation_diagram.png)

IfcRelSequence seems necessary if tasks have overlapping duration. 

With all that said, for the assumed construction sequence needed, dates, times, and duration is general unknown. My assumption is that the task long description that says fill blackouts with grout within 24 hours of cutting strands sufficiently fulfills the requirement

**Reply by MatthiasWeise** [reply]
_IfcRelNests_ defines an order of subtasks, but this order was for visualization purposes only as for instance needed in a Gant chart. 
Of course, it is up to interpretation whether you want to derive a logical task dependency from this order but this solution is limited to a sequential order and does not support definition of tasks that can be carried out in parallel. We may use _IfcRelSequence_ in such cases but such exceptions will make software implementation more complicated than just following one generic rule.  

This explains the following sentences in the IFC documentation: 
_The task order information that is used for viewing purposes is derived from the order defined by the IfcRelNests relationship and thus is independent of the logical task order defined through IfcRelSequence. The hierarchy and order defined through IfcRelNests enables to order the tasks in a tree view or list view structure._

**Reply by RickBrice** [reply]
> Passes all basic checks, apart from: 13s6piTaH4duITiwjqo2Ah IfcTask is not declared.

Can you explain this a bit more?

~~~
#109=IFCTASK('13s6piTaH4duITiwjqo2Ah',$,'Construct Bridge','Assumed construction sequence',$,'0','Provided per requirements of AASHTO LRFD BDS 2.5.3',$,$,.T.,$,$,.CONSTRUCTION.);
#110=IFCRELASSIGNSTOCONTROL('3c$PgwL6LAagZBcPBxkpfr',$,'Assumed Construction Sequence Work Schedule Tasks',$,(#109),$,#107);
#107=IFCWORKSCHEDULE('0SPT5ZPkXC7BPkxCN__1ZF',$,'Assumed Construction Sequence',$,$,$,'2025-01-14T15:27:29Z',$,'Satisfies the requirements of AASHTO LRFD BDS 2.5.3',$,$,'2025-01-14T15:27:29Z',$,.PLANNED.);
~~~

The summary task is assigned to the IfcWorkSchedule per diagrams in IfcWorkSchedule and IfcTask

![image](https://github.com/user-attachments/assets/29cc2fa3-5750-4cf5-94f5-a3fbe5065b9c)


IfcTask would only be IfcRelDeclares with IfcProject if it were a root-level task. "If an assigned [IfcTask](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcTask.htm) is a root-level task, such task must be declared on the [IfcProject](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcProject.htm) using the [IfcRelDeclares](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcRelDeclares.htm) relationship."

**Reply by SergejMuhic** [reply]
Alright, this seems reasonable of course. 😄 I modified the check a bit according to the diagram above. It expects an `IfcTask` declared on `IfcProject`.
```IfcProject - IfcRelDeclares - IfcTask (summary) - IfcRelNests - [Subtasks IfcTask, IfcTask, etc]```

Let me modify this accordingly. I have now created a Project root to deal with the declarations specifically (according to your diagram) and will change the general check of declarations back to do it more generally, then this reported error:
"13s6piTaH4duITiwjqo2Ah IfcTask is not declared."
will not pop up anymore.

**Reply by RickBrice** [reply]
I don't see declaring a root level task on IfcProject any more or less valid than having IfcWorkSchedule - IfcRelAssignsToControl - IfcTask. Through discussions on OSArch forums, I was of the understanding that software would generally support the IfcWorkSchedule approach better. That's the only reason I went down that path.

**Reply by SergejMuhic** [reply]
"I don't see declaring a root level task on IfcProject any more or less valid than having IfcWorkSchedule - IfcRelAssignsToControl - IfcTask." - yes, of course. But this was not the issue. The check was looking for IfcTask being declared or nested, but if it can also be assigned to control, this test passes.

I have modified this now to not be specific for IfcTask having to be declared (or decomposed, or nested ...) but also allowing assignment (as was originally). For specifics, MVD definitions make sense.

**Reply by SergejMuhic** [reply]
After all the discussions and some back and forth with @MatthiasWeise on Skype I have now used the following approach on `IfcTask:`
![image](https://github.com/user-attachments/assets/103579f6-d297-4f8f-a254-85ed8725441c)
and
![image](https://github.com/user-attachments/assets/ef5ac10c-5a97-4073-b176-f3917e2de167)

The tests are successful. I can still of course also provide the tests on Nesting and Project declarations separately. Can also go down the `IfcWorkPlan` and `IfcWorkSchedule` route.

And one the other question would be whether to setup the `IfcTask.LongDescription` or my proposal with `IfcTaskTime` [here](https://github.com/jwouellette/TPF-5_372-Unit_Test_Suite/discussions/16#discussioncomment-11880545)?

Results:
![image](https://github.com/user-attachments/assets/0c772f69-9805-4d09-8cbb-387bbb67a4e7)

**Reply by SergejMuhic** [reply]
And played a bit more and produced a documentation showcase:
![image](https://github.com/user-attachments/assets/ea0cd3d3-d2c5-472d-a709-e2f3ac0fda5f)

You see I have some formatting going wrong there but I did not want to spend time with aesthetics. 😄 The css is obviously also unchanged but bear with me.

**Comment by SergejMuhic** [comment]
Your proposal @RickBrice can easily be encoded into usages on IfcTask and the relevant IfcProduct subtypes. IfcProject with the Project Declaration usage:
![image](https://github.com/user-attachments/assets/a6e7285a-120a-46fe-9edc-69b5dd7a5428)
already supports the first part.

This is the purpose of MVDs, defining these specific usages and making them public in a computer and human readable format.

I can offer to model this and derive a checker app from it. I can also provide the UI to fill in specific requirements. All can be done, but as you say not in the domain of IDS.

I can also understand that this would not be wanted in the spec due to opinions and excluding technologies, but I stick to my convictions that using the data model and the data structures according to their design purpose is the way to go. Otherwise, we could just use tables anyway. What would then the benefit of using IFC be? Exclusion because it is not implemented is not a good argument. Progress does not come from a place of comfort.

Regarding improvements, to my knowledge bSI has sacked their bSI process and are working on a new one:
![image](https://github.com/user-attachments/assets/cc4d1b48-62c6-4d57-ac92-d390b1a2bf13)
https://www.buildingsmart.org/about/bsi-process/

I would just talk to Evandro directly to include this in the AbV.

**Reply by RickBrice** [reply]
@SergejMuhic 

> I would just talk to Evandro directly to include this in the AbV.

We have already tried this.

I would like to see @jwouellette, or someone from the TPF team set up a demonstration and discussion of your work. Generating checking apps from the MVD (mvdxml) seems pretty useful to me (though I'm probably a bit naive). Not sure if it complements IDS or is a competing approach and possible replacement for IDS. At any rate, the folks involved in our efforts at the State DOT side should know about this approach.

**Reply by MatthiasWeise** [reply]
Just to clarify from my point of view: 
IDS can be considered as a supplement, but does not replace an appropriate technical agreement on the subset of the IFC schema to be used and the way to implement IFC, if additional implementer agreements are necessary. 

While an MVD defines how to use IFC, IDS is purely focused on required content (what information is needed). By design, IDS hides implementation details. Furthermore, current version 1.0 is limited to the agreed scope, meaning to support requirements that can be delivered by properties, attribues, existance of material, classification and basic part of requirements. 
Requirements for (task) sequencing, assignment to product, etc. as needed in this context are out of scope.

**Reply by SergejMuhic** [reply]
'I would like to see @jwouellette, or someone from the TPF team set up a demonstration and discussion of your work.' I have talked a couple of times with @jwouellette about it but we have so many topics that we always deviate. Maybe we setup a dedicated meeting and focus? But then at least one of you two @RickBrice and @MatthiasWeise (preferably even both) should be present to keep the focus.

**Comment by MatthiasWeise** [comment]
A good summary of the decisions to be made by all parties involved! 

Some additional questions that come into my mind, in particular related to the example given by @RickBrice :

- Are we going to model temporary elements (you mentioned a few like temporary bracing, expanded polystyrene, etc.)? If using _IfcTask_ we do not need an element that is linked via _IfcRelAssignsToProduct_. So, unless we plan to model temporary elements it would be an another argument for using _IfcTask_ instead of custom properties. 
- Are we happy how elements are split in our design, does it work for task sequencing? Seeing definitions like ‘top flange of a beam’, it shows that we may not have the same granularity of elements between design and construction. This is somehow to be expected for the construction planning, but the question is whether our design is good enough for describing the sequence of tasks.
- There are additional time dependencies such as "within 24 hours ..". While there is a way to define a time lag between tasks in _IfcRelSequence_, I do not know a way to model maximum delays.

**Reply by RickBrice** [reply]
@MatthiasWeise - all good questions

> Are we going to model temporary elements (you mentioned a few like temporary bracing, expanded polystyrene, etc.)? If using IfcTask we do not need an element that is linked via IfcRelAssignsToProduct. So, unless we plan to model temporary elements it would be an another argument for using IfcTask instead of custom properties

That is a good LOD question. Generally, at my State DOT, we don't design the contractor's temporary works. We just say they need to use temporary bracing to stabilize girders during construction.

> Are we happy how elements are split in our design, does it work for task sequencing? Seeing definitions like ‘top flange of a beam’, it shows that we may not have the same granularity of elements between design and construction. This is somehow to be expected for the construction planning, but the question is whether our design is good enough for describing the sequence of tasks.

Can't we rely on some implied domain knowledge here? All parties involved should know what `top flange of beam` means, what it looks like, and where it is located.

> There are additional time dependencies such as "within 24 hours ..". While there is a way to define a time lag between tasks in IfcRelSequence, I do not know a way to model maximum delays.

This is also a LOD question. The contract requirements are that the blockouts are grouted within 24 hours. Different owners may have other requirements. This can be modeled in a schedule (or maybe not if IFC doesn't support maximum lag time) or stated in the long description of a IfcTask.

**Reply by MatthiasWeise** [reply]
Ok, then a link to a specific element of the bridge seems to be an optional information.

Regarding maximum lag time: 
I do not know how to best deal with this information if you want to use IfcTasks. It is basically an information of IfcRelSequence, which however cannot have own properties ;-). 
On solution/work-around might be to use an additional IfcEvent in between both tasks, which automatically triggers the next task after 24 h.

**Reply by RickBrice** [reply]
My approach was to use IfcTask.LongDescription

`#117=IFCTASK('2s6MTzjWv7tRUWD2YbSZZN',$,'Fill blockouts',$,$,'1.4.3','Within 24 hours of cutting the temporary strands, fill the blockouts with a grout conforming to Std. Spec. 9-20.3(2). Remove all moisture in blockouts prior to filling them with grout.',$,$,.T.,$,$,.INSTALLATION.);`

**Reply by SergejMuhic** [reply]
I might not understand the issue in question but is the attribute `IfcTask.TaskTime` not fit for this purpose?

https://bsi-infraroom.github.io/IFC-Documentation-Tunnel/4_4_0_0/general/HTML/link/ifctasktime.htm

A combination of:
`IfcTaskTime.ScheduleDuration`
and
`IfcTaskTime.IsCritical`

should do the trick, shouldn't it?

**Reply by MatthiasWeise** [reply]
The discussion is about time-dependencies between tasks, ideally attached to _IfcRelSequence_. We may use a combination of finish + early/late start time information to describe max. time lags, but that seems to be another work-around as we would have to use _IfcDateTime_, which identifies a particular point in time.
